elasticsearch/src/main/java/org/apache/calcite/adapter/elasticsearch/ElasticsearchTable.java

```diff
@@ -87,37 +126,211 @@ private static ElasticsearchVersion detectVersion(RestClient client, ObjectMappe
+  private Enumerable<Object> aggregate(List<String> ops,
+      List<Map.Entry<String, Class>> fields,
+      List<Map.Entry<String, RelFieldCollation.Direction>> sort,
+      List<String> groupBy,
+      List<Map.Entry<String, String>> aggregations,
+      Long offset, Long fetch) throws IOException {
+
+    if (aggregations.isEmpty()) {
+      throw new IllegalArgumentException("Missing Aggregations");
+    }
+
+    if (!groupBy.isEmpty() && offset != null) {
+      String message = "Currently ES doesn't support generic pagination "
+          + "with aggregations. You can still use LIMIT keyword (without OFFSET). "
+          + "For more details see https://github.com/elastic/elasticsearch/issues/4915";
+      throw new IllegalStateException(message);
+    }
+
+    final ObjectNode query = mapper.createObjectNode();
+
+    // manually parse into JSON from previously concatenated strings
+    query.setAll((ObjectNode) mapper.readTree("{" + Util.toString(ops, "", ", ", "") + "}"));
+
+    // remove / override attributes which are not applicable to aggregations
+    query.put("_source", false);
+    query.put("size", 0);
+    query.remove("script_fields");
+
+    // allows to detect aggregation for count(*)
+    final Predicate<Map.Entry<String, String>> isCountStar = e -> e.getValue()
+            .contains("\"" + ElasticsearchConstants.ID + "\"");
+
+    // list of expressions which are count(*)
+    final Set<String> countAll = aggregations.stream()
+            .filter(isCountStar)
+        .map(Map.Entry::getKey).collect(Collectors.toSet());
+
+    final Map<String, String> fieldMap = new HashMap<>();
+
+    // due to ES aggregation format. fields in "order by" clause should go first
+    // if "order by" is missing. order in "group by" is un-important
+    final Set<String> orderedGroupBy = new LinkedHashSet<>();
+    orderedGroupBy.addAll(sort.stream().map(Map.Entry::getKey).collect(Collectors.toList()));
+    orderedGroupBy.addAll(groupBy);
+
+    // construct nested aggregations node(s)
+    ObjectNode parent = query.with(AGGREGATIONS);
+    for (String name: orderedGroupBy) {
+      final String aggName = "g_" + name;
+      fieldMap.put(aggName, name);
+
+      final ObjectNode section = parent.with(aggName);
+      final ObjectNode terms = section.with("terms");
+      terms.put("field", name);
+      terms.set("missing", ElasticsearchJson.MISSING_VALUE); // expose missing terms
+
+      if (fetch != null) {
+        terms.put("size", fetch);
+      }
+
+      sort.stream().filter(e -> e.getKey().equals(name)).findAny().ifPresent(s -> {
+        terms.with("order").put("_key", s.getValue().isDescending() ? "desc" : "asc");
+      });
+
+      parent = section.with(AGGREGATIONS);
+    }
+
+    // simple version for queries like "select count(*), max(col1) from table" (without GROUP BY)
+    if (!aggregations.stream().allMatch(isCountStar) || !groupBy.isEmpty()) {
```

Comment:
The change of the order of disjuncts was intentional to avoid the rather costly stream operation if there is a group by. Please swap the two disjuncts.

Reply:
OK. Now I see what you mean. Fixed